test/integration/examples/apiserver_test.go

```diff
@@ -90,52 +92,49 @@ func TestAggregatedAPIServer(t *testing.T) {
 
 	kubeClientConfigValue := atomic.Value{}
 	go func() {
-		for {
-			// always get a fresh port in case something claimed the old one
-			kubePort, err := framework.FindFreeLocalPort()
-			if err != nil {
-				t.Fatal(err)
-			}
-
-			kubeAPIServerOptions := options.NewServerRunOptions()
-			kubeAPIServerOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
-			kubeAPIServerOptions.SecureServing.BindPort = kubePort
-			kubeAPIServerOptions.SecureServing.ServerCert.CertDirectory = certDir
-			kubeAPIServerOptions.InsecureServing.BindPort = 0
-			kubeAPIServerOptions.Etcd.StorageConfig.ServerList = []string{framework.GetEtcdURL()}
-			kubeAPIServerOptions.ServiceClusterIPRange = *defaultServiceClusterIPRange
-			kubeAPIServerOptions.Authentication.RequestHeader.UsernameHeaders = []string{"X-Remote-User"}
-			kubeAPIServerOptions.Authentication.RequestHeader.GroupHeaders = []string{"X-Remote-Group"}
-			kubeAPIServerOptions.Authentication.RequestHeader.ExtraHeaderPrefixes = []string{"X-Remote-Extra-"}
-			kubeAPIServerOptions.Authentication.RequestHeader.AllowedNames = []string{"kube-aggregator"}
-			kubeAPIServerOptions.Authentication.RequestHeader.ClientCAFile = proxyCACertFile.Name()
-			kubeAPIServerOptions.Authentication.ClientCert.ClientCA = clientCACertFile.Name()
-			kubeAPIServerOptions.Authorization.Mode = "RBAC"
-
-			tunneler, proxyTransport, err := app.CreateNodeDialer(kubeAPIServerOptions)
-			if err != nil {
-				t.Fatal(err)
-			}
-			kubeAPIServerConfig, sharedInformers, versionedInformers, _, _, err := app.CreateKubeAPIServerConfig(kubeAPIServerOptions, tunneler, proxyTransport)
-			if err != nil {
-				t.Fatal(err)
-			}
-			// Adjust the loopback config for external use (external server name and CA)
-			kubeAPIServerClientConfig := rest.CopyConfig(kubeAPIServerConfig.GenericConfig.LoopbackClientConfig)
-			kubeAPIServerClientConfig.CAFile = path.Join(certDir, "apiserver.crt")
-			kubeAPIServerClientConfig.CAData = nil
-			kubeAPIServerClientConfig.ServerName = ""
-			kubeClientConfigValue.Store(kubeAPIServerClientConfig)
-
-			kubeAPIServer, err := app.CreateKubeAPIServer(kubeAPIServerConfig, genericapiserver.EmptyDelegate, sharedInformers, versionedInformers)
-			if err != nil {
-				t.Fatal(err)
-			}
-
-			if err := kubeAPIServer.GenericAPIServer.PrepareRun().Run(wait.NeverStop); err != nil {
-				t.Log(err)
-			}
-			time.Sleep(100 * time.Millisecond)
+		listener, port, err := genericapiserveroptions.CreateListener("tcp", "127.0.0.1:0")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		kubeAPIServerOptions := options.NewServerRunOptions()
+		kubeAPIServerOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
```

Comment:
wasn't bind address and port automatic when the listener is set?

Reply:
Yes,  when listener is set.  BindAddress  and Port  no need to set. 
I set here to keep consistency.

Reply:
Actually we donot need `BindPort` and `BindAddress`.

Reply:
I see a nit bug, and will open a pr to fix it in `SecureServingOptions.ApplyTo`